graph.go

```diff
@@ -4,193 +4,166 @@
 
 package graph
 
-// All a node needs to do is identify itself. This allows the user to pass in nodes more
-// interesting than an int, but also allow us to reap the benefits of having a map-storable,
-// comparable type.
+import "math"
+
+// Node is a graph node. It returns a graph-unique integer ID.
 type Node interface {
 	ID() int
 }
 
-// Allows edges to do something more interesting that just be a group of nodes. While the methods
-// are called From and To, they are not considered directed unless the given interface specifies
-// otherwise.
+// Edge is a graph edge. In directed graphs, the direction of the
+// edge is given from -> to, otherwise the edge is semantically
+// unordered.
 type Edge interface {
 	From() Node
 	To() Node
 }
 
-// A Graph implements the behavior of an undirected graph.
-//
-// All methods in Graph are implicitly undirected. Graph algorithms that care about directionality
-// will intelligently choose the DirectedGraph behavior if that interface is also implemented,
-// even if the function itself only takes in a Graph (or a super-interface of graph).
+// Graph is a generalized graph.
 type Graph interface {
-	// NodeExists returns true when node is currently in the graph.
-	NodeExists(Node) bool
-
-	// NodeList returns a list of all nodes in no particular order, useful for
-	// determining things like if a graph is fully connected. The caller is
-	// free to modify this list. Implementations should construct a new list
-	// and not return internal representation.
-	NodeList() []Node
-
-	// Neighbors returns all nodes connected by any edge to this node.
-	Neighbors(Node) []Node
-
-	// EdgeBetween returns an edge between node and neighbor such that
-	// From is one argument and To is the other. If no
-	// such edge exists, this function returns nil.
-	EdgeBetween(node, neighbor Node) Edge
-}
+	// Has returns whether the node exists within the graph.
```

Comment:
The problem with if is that it is heavily overloaded. It's not possible to determine whether "if" in a sentence is "if" or "iff". Worse, it defines a behaviour that is arguably not true. It is possible to legitimately interpret that sentence as Has returns a value true if the node exists, but also if it doesn't; or it panics, or some other undefined behaviour. "Whether" says that it returns a boolean that is defined entirely by the state of presence of the node in the graph. This is probably the reason for the relatively recent purge of the use of "if" replaced by "whether" in the standard library.